### PR TITLE
Implement pinned memory

### PR DESCRIPTION
**Pull request by okuta**
Please merge #1709 before this PR.

### CODE REVIEW

**Comment by rezoo** [comment]
# First impression
It seems that this PR is attractive; the disadvantage of this PR is the increase of allocation time, but in many cases where we use cupy for large-scale computation and need to allocate memory more than one megabyte, it significantly reduces the total time.

I will read it in detail :)

**Comment by rezoo** [comment]
Are there any reasons why ``setStream`` and ``getStream`` do not use ``with nogil:`` statement?

File: cupy/cuda/pinned_memory.pyx [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    cpdef free(self, size_t ptr, Py_ssize_t size):
+        cdef list free
+        cdef PinnedMemory mem
+        mem = self._in_use.pop(ptr, None)
+        if mem is None:
+            raise RuntimeError('Cannot free out-of-pool memory')
+        free = self._free[size]
+        free.append(mem)
+
+    cpdef free_all_free(self):
```
Comment (rezoo) [inline cupy/cuda/pinned_memory.pyx]: ``free_all_free`` is difficult to understand; how about replacing it with ``free_all_blocks``?

File: cupy/cuda/pinned_memory.pyx [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+            size are all in use.
+
+    """
+
+    def __init__(self, allocator=_malloc):
+        self._in_use = {}
+        self._free = collections.defaultdict(list)
+        self._alloc = allocator
+        self._weakref = weakref.ref(self)
+        self._allocation_unit_size = 512
```
Comment (rezoo) [inline cupy/cuda/pinned_memory.pyx]: Is it valid that ``allocation_unit_size`` is always 512 bytes regardless of devices?

File: cupy/cuda/pinned_memory.pyx [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+        """
+        pool = self.pool()
+        if pool and self.ptr != 0:
+            pool.free(self.ptr, self.size)
+        self.ptr = 0
+        self.size = 0
+
+
+cdef class PinnedMemoryPool:
```
Comment (rezoo) [inline cupy/cuda/pinned_memory.pyx]: I am afraid of the case where we have to allocate a lot of buffers having different sizes, and makes the memory pool bloat. Such a variety of the buffers are sometimes occurred in some fields (e.g., NLP that have to allocate variety of arrays). Did the chainer team agree on that?

File: cupy/cuda/pinned_memory.pyx [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        size = (((size + unit - 1) // unit) * unit)
+        free = self._free[size]
+        if free:
+            mem = free.pop()
+        else:
+            try:
+                mem = self._alloc(size).mem
+            except runtime.CUDARuntimeError as e:
+                if e.status != 2:
+                    raise
```
Comment (okuta) [inline cupy/cuda/pinned_memory.pyx]: This is re-raise statement.
Why do we use explicitly raise statement?

File: cupy/cuda/pinned_memory.pyx (hunk 3 above)
Comment (okuta) [inline cupy/cuda/pinned_memory.pyx]: The memory pool of CDUA memory has same problem.
When we find better memory pooling strategy, we should change behavior.

File: cupy/cuda/pinned_memory.pyx (hunk 2 above)
Comment (okuta) [inline cupy/cuda/pinned_memory.pyx]: Not always.
But, I think that this configuration is not too bad.

File: cupy/cuda/pinned_memory.pyx (hunk 4 above)
Comment (rezoo) [inline cupy/cuda/pinned_memory.pyx]: Ah sorry I missed it, but I have another questions:
why you ignore cudaErrorMemoryAllocation? If you have reasons, please describe them as a comment.

File: cupy/cuda/pinned_memory.pyx (hunk 3 above)
Comment (rezoo) [inline cupy/cuda/pinned_memory.pyx]: > The memory pool of CDUA memory has same problem.
> When we find better memory pooling strategy, we should change behavior.
I see. I summarize my conclusion in the below.

File: cupy/cuda/pinned_memory.pyx (hunk 2 above)
Comment (okuta) [inline cupy/cuda/pinned_memory.pyx]: I checked in my environment. 512 is the looks good.
I do not know about usual case in the cuda programming.

File: cupy/cuda/pinned_memory.pyx (hunk 4 above)
Comment (okuta) [inline cupy/cuda/pinned_memory.pyx]: I want to call `free_all_blocks` for retry allocation.

File: cupy/cuda/pinned_memory.pyx [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+    def __add__(x, y):
+        """Adds an offset to the pointer."""
+        cdef PinnedMemoryPointer self
+        cdef Py_ssize_t offset
+        if isinstance(x, PinnedMemoryPointer):
+            self = x
+            offset = <Py_ssize_t?>y
+        else:
+            self = <PinnedMemoryPointer?>y
+            offset = <Py_ssize_t?>x
```
Comment (rezoo) [inline cupy/cuda/pinned_memory.pyx]: I see.

File: cupy/cuda/pinned_memory.pyx (hunk 2 above)
Comment (rezoo) [inline cupy/cuda/pinned_memory.pyx]: OK I understand what you say. If the memory pool fails to allocate, it calls ``free_all_blocks`` and release unused memory.

File: cupy/cuda/pinned_memory.pyx (hunk 4 above)
Comment (rezoo) [inline cupy/cuda/pinned_memory.pyx]: I see.

**Comment by rezoo** [comment]
In conclusion, I think that this code itself looks good and is OK to merge.
My feeling is that it is very hard to read the whole comment due to the lack of the comments.
I think it is good to describe more comments regarding the objective of the code and its meaning.
It is important for the other persons to maintain your code, fix bugs, and make it improve.

As this is a large PR, I cannot decide whether it is OK to merge. I asked few guys to read the whole code. After the guys agree on that, I would like to merge this attractive PR.

File: cupy/cuda/pinned_memory.pyx (hunk 4 above)
Comment (unnonouno) [inline cupy/cuda/pinned_memory.pyx]: What is this magic number `2`?

File: cupy/cuda/stream.py [hunk 6]
```diff
@@ -161,4 +161,7 @@ def wait_event(self, event):
             event (cupy.cuda.Event): CUDA event.
 
         """
-        runtime.streamWaitEvent(self.ptr, event)
+        runtime.streamWaitEvent(self.ptr, event.ptr)
+
+
+NullStream = Stream(null=True)
```
Comment (unnonouno) [inline cupy/cuda/stream.py]: Is this an instance? If so, CameCase feels confusing. It looks type name.

**Comment by gwtnb** [comment]
```
======================================================================
ERROR: test_free_all_free (test_pinned_memory.TestSingleDeviceMemoryPool)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/chainer/tests/cupy_tests/cuda_tests/test_pinned_memory.py", line 102, in test_free_all_free
    self.pool.free_all_free()
AttributeError: 'cupy.cuda.pinned_memory.PinnedMemoryPool' object has no attribute 'free_all_free'

======================================================================
ERROR: test_free_all_free2 (test_pinned_memory.TestSingleDeviceMemoryPool)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/chainer/tests/cupy_tests/cuda_tests/test_pinned_memory.py", line 113, in test_free_all_free2
    self.pool.free_all_free()
AttributeError: 'cupy.cuda.pinned_memory.PinnedMemoryPool' object has no attribute 'free_all_free'

======================================================================
ERROR: test_free_all_free_without_malloc (test_pinned_memory.TestSingleDeviceMemoryPool)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/chainer/tests/cupy_tests/cuda_tests/test_pinned_memory.py", line 128, in test_free_all_free_without_malloc
    self.pool.free_all_free()
AttributeError: 'cupy.cuda.pinned_memory.PinnedMemoryPool' object has no attribute 'free_all_free'

----------------------------------------------------------------------
```

**Comment by rezoo** [comment]
@gwtnb @unnonouno @okuta  I manually confirmed that the current branch surely resolves the above issues except for ``NullStream``. 

i) Can you continue the review?
ii) In terms of ``NullStream``, I think either is fine. Please make an agreement.

**Comment by rezoo** [comment]
Good. Then I would like to merge this one 👍

**Pull request merged**